### DISCUSSION TITLE
Unable to use Previewer with Dependency Injection

### DISCUSSION THREAD

**Discussion by laolarou726** [AvaloniaUI/Avalonia#13743] [Q&A] [answered]
## Describe the bug

Currently, AvaloniaUI's previewer can not work well when I use `Microsoft.Extensions.DependencyInjection` to inject all the types into the constructor. Previewer will show the following messages:

```
13:58:10.093 [Information] 45904 Connection initialized
13:58:12.131 [Error] 45904 UpdateXamlResult error
System.Xaml.XamlException: “Unable to find public constructor for type xxx:xxxx.xxxx.xxxViewModel() Line 26, position 10.”
```

## To Reproduce

1. Create a simple view with constructor and parameters
2. Build the project

## Expected behavior

The preview window should have the rendered controls instead of errors

## Environment 

- OS: Windows 11 23H2
- Avalonia-Version: 11.0.5

**Comment by timunie** [comment]
Try if(Design.IsDesignTime)

**Comment by laolarou726** [comment]
> Try if(Design.IsDesignTime)

Is there a way to detect the previewer using the compile symbol? I don't want those codes to be present in the production publish.

**Reply by timunie** [reply]
Then you need to find another solution. Maybe `#IF DEBUG` can help here, but I never tested this with the designer.

**Comment by maxkatz6** [comment]
It doesn't look like an issue. The error message is pretty descriptive and specific.
Can you tell what exactly are you trying to do? I.e. specific XAML code.

**Reply by laolarou726** [reply]
All the preview views in my Project are broken because I'm using MSDI to inject the dependencies, and it seems that Avalonia failed to create those view and view model instances in the design mode. So the previewer has no content but error message.

**Reply by maxkatz6** [reply]
MSDI shouldn't be broken because of the previewer. But you might have used it wrong. Or, maybe, set it up in the Main method which is not executed by the previewer either.

**Reply by maxkatz6** [reply]
See https://github.com/AvaloniaUI/Avalonia.Samples/pull/64

**Reply by nixtar** [reply]
I ran into this also when I first started using Avalonia.

I normally like to create my WPF VMs like this: (Coming from an ASP.NET Core background)

```C#
public class MainViewModel : ObservableObject
{
    private readonly AppConfig _appConfig;

    public MainViewModel(AppConfig appConfig)
    {
        _appConfig = appConfig;
    }
}
```

Since using avalonia I've had to switch to using:

```C#
public class MainViewModel : ObservableObject
{
    private readonly AppConfig _appConfig;

    public MainViewModel()
    {
        _appConfig = Ioc.Default.GetRequiredService<AppConfig >();
    }
}
```

This example is using the static Ioc extensions provided in the CommunityToolkit.Mvvm.DependencyInjection namespace.

Looking at [src/Avalonia.Samples/MVVM/MsExtensionsHostingSample/ViewModels/DesignData.cs](https://github.com/AvaloniaUI/Avalonia.Samples/pull/64/files#diff-e18889919c51bee70fe264877f7dfb8a9bc1a2bc828f351c092df577f2a6a54d) from https://github.com/AvaloniaUI/Avalonia.Samples/pull/64 is interesting. If it works the way I think it does it could be a neat solution to this also.

**Reply by laolarou726** [reply]
Just took a look at the sample, and it still requires making all the View's constructors no parameter at all. I was wondering if there is a way to keep using the constructor with parameters for the Views and the preview for controls.

**Reply by maxkatz6** [reply]
> and it still requires making all the View's constructors no parameter at all

Yes. Previewer wouldn't know how to create your special type. That's why most of the type IoC works with view models and not views.

**Reply by maxkatz6** [reply]
> Since using avalonia I've had to switch to using:

You can use parameters in view models with no issues. DesignData is one of the possible solutions, yes - my favorite.

**Reply by laolarou726** [reply]
I'm pretty sure WPF's previewer can use parameterized ctor on view without causing any trouble. Any ideas for migrating this feature to AvaloniaUI?

**Reply by nixtar** [reply]
> I'm pretty sure WPF's previewer can use parameterized ctor on view without causing any trouble. Any ideas for migrating this feature to AvaloniaUI?

I think the confusion here is that WPF doesn't have "MVVM" templates and by default there is no `d:DataContext` set.

If you wire up a WPF app with the same thing as Avalonia's MVVM templates, with `d:DataContext=` pointing at a view model with a param you get this:
![image](https://github.com/AvaloniaUI/Avalonia/assets/2175630/79f11001-161e-4234-a1ad-01fa561c6609)

Remove the param and you get the same behaviour as the MVVM Avalonia templates:
![image](https://github.com/AvaloniaUI/Avalonia/assets/2175630/458e36ea-25a1-44b2-83df-bfe6b2bd1da6)

Another alternative is to remove the `Design.DataContext` block from your Avalonia view but keep `x:DataType` (assuming using compiled bindings).

This shouldn't need to construct the VM in the designer, but the preview won't display the bindings in the designer (Just like WPF without `d:DataContext`).

**Reply by nixtar** [reply]
Ahh sorry, I missed the part about this being about Views and not just ViewModels...

**Reply by thevortexcloud** [reply]
I will add that the previewer not easily working with DI and view models is one of the main reasons I don't bother using the previewer.  Most of the time it just crashes with null reference exceptions due to me trying to use ReactiveUI stuff in code behind that references the view model. Being able to inject a bunch of fake test services while designing layouts would be very useful. The XAML design time stuff is fairy verbose and annoying to maintain (and somewhat defeats the point of using DI in the first place). But I am not sure what a good solution here would be.

That said: I am fine with using the XAML only view anyway.

**Reply by timunie** [reply]
`if(Desing.IsDesignTime)` is a valid solution for me. Maybe one can also wrap it in `# IF DEBUG` but I haven't tested it.

**Reply by thevortexcloud** [reply]
I don't mind doing that when setting up the DI container if I need to. But as far as I can tell it's not possible to get that to work. But doing it in _every_ view with a service dependency is not really worth the effort and not much different from using the XAML design time stuff. I suppose I could use use a locator call only on debug builds to resolve the view model with the test dependencies, but I am still not a fan of doing that.

**Reply by timunie** [reply]
well if anyone has a better idea, we are open to it.

**Reply by thevortexcloud** [reply]
Hmm. The simplest option I can think of, of the top of my head which I suspect may be somewhat contentious would be to allow for Avalonia to resolve view models from it's own locator, either on an opt in or opt out basis. Probably with some sort of property on the view to control it (which may have other problems if it needs to instantiate the view to see the property value). But this then requires people to interact with Splat in some way, either to directly register things or to make it use another DI framework. In which case it would probably be a good idea to provide some sort of abstraction over it in the event Avalonia ever decides to move to a different internal DI.

But as I said, I can't think of any particularly great generic ideas for doing this, which is why I have not really complained that much about it.

**Reply by jp2masa** [reply, marked as answer]
In my projects I usually create a dummy DI container and then do something like `Design.DataContext={markup:DesignResolve {x:Type vm:MainViewModel}}`, with a custom markup extension that resolves view models from the dummy container.

I believe it could be simpler, for example by creating a base class for views or overriding the default value for `Design.DataContextProperty`.

Additionally, if you want to use DI with views as well, I think overriding the default value for `Design.PreviewWithProperty` would work.

**Reply by thevortexcloud** [reply]
I just tried that and it works. I had not thought about using a markup extension. Very nice. Is there a chance we can get something like this added to Avalonia itself to avoid having to copy this code everywhere I need it? 

On a related note: is there a way to get the `x:DataType` if I have a reference to the control? It would allow me to make this even simpler.

**Reply by timunie** [reply]
I don't think this is well suited for vanilla Avalonia. But it *may* make it into Avalonia.Labs or can be added to a community third party package which can be easily added via nuget.

**Reply by thevortexcloud** [reply]
Fair enough.

**Reply by leowest77** [reply]
@jp2masa Hi, I am in similar issue using DI and ViewLocator could you provide an example of how to use the options you provided? Apparently I cannot figure it out myself. But I really would love to get ride of this issue `Cannot compile Build method. Parameters mismatch.Type needs to have a parameterless ctor or a ctor with a single IServiceProvider argument.Or x:Arguments directive with matching arguments needs to be set`  even if I see the a parameterless ctor it still does no like it.

**Reply by thevortexcloud** [reply]
Here is my implementation of it. It's fairly basic though.

```csharp
 public class DesignLocator : MarkupExtension {
        #region Public properties
        public Type Type { get; init; }
        #endregion

        #region Public methods
        public override object ProvideValue(IServiceProvider serviceProvider) {
            if (!Design.IsDesignMode) {
                throw new InvalidOperationException();
            }

            return App.Locator.GetRequiredService(this.Type);
        }
        #endregion
    }
```

And in the App.xaml.cs file I have put a static locator like this:

```csharp
        #region Public properties
        public static ServiceProvider Locator { get; private set; }
        #endregion

        #region Public methods
        public override void Initialize() {
            AvaloniaXamlLoader.Load(this);
        }

        public override void OnFrameworkInitializationCompleted() {
            var collection = new ServiceCollection();
            //Register services
            if (Design.IsDesignMode) {
                collection.RegisterDesignTimeServices();
            } else {
                collection.RegisterRealServices();
            }


            var services = collection.BuildServiceProvider();
            Locator = services;
            if (this.ApplicationLifetime is IClassicDesktopStyleApplicationLifetime desktop) {
                desktop.MainWindow = new MainWindow {
                    DataContext = services.GetRequiredService<MainWindowViewModel>()
                };
            }

            base.OnFrameworkInitializationCompleted();
        }
        #endregion
```

The key thing this does it will use the DI container to instantiate the instance of the view model, rather than Avalonia itself trying to do it. You should also make sure you register fake services (like I am doing) or you may end up triggering things you likely don't want to when looking at the previewer.

**Reply by leowest77** [reply]
Appreciate the answer ~~but I might be missing something to understand how DesignLocator is triggered here...~~

~~Or would that be this inside the UserControl `Design.DataContext={markup:DesignResolve {x:Type vm:MainViewModel}}` except I need to rename Resolve to Locator? also is markup being used as the namespace?~~

~~I think it seems to be this but now it complains of~~

```
Unable to find public constructor for type DesignLocator(System.Private.CoreLib:System.Type)
```
~~I did give it a~~
```
        public DesignLocator(Type type)
        {
            Type = type;
        }
```
~~but still complains.~~

OK, I have no idea what happened but now it accepted it.. I even restarted vs just to be sure its not complaining anymore hehe. Thanks for the help.

**Reply by zangetsudo** [reply]
if (Design.IsDesignMode) {
                collection.RegisterDesignTimeServices();
            } else {
                collection.RegisterRealServices();
            }

cannot find RegisterDesignTimeServices and RegisterRealServices. this code can't be compiled.

**Reply by thevortexcloud** [reply]
That's because it's a custom extension method. Use the sample code I wrote in the DI tutorial to get you started:

https://docs.avaloniaui.net/docs/app-development/dependency-injection#step-2-add-servicecollectionextensions

**Reply by zangetsudo** [reply]
> That's because it's a custom extension method. Use the sample code I wrote in the DI tutorial to get you started:
> 
> https://docs.avaloniaui.net/docs/app-development/dependency-injection#step-2-add-servicecollectionextensions

sorry, still cannot find anything related to this code. i’ve seen this post carefully before.
thanks anyway this guide has helped me to understand DI very much.

**Reply by thevortexcloud** [reply]
The `RegisterDesignTimeServices` and `RegisterRealServices` are custom methods specific to my app. The specific implementation is not important for the example (and will look very different for every app), which is why I did not bother showing it. The link I gave you shows you how to make your own custom extension methods to register services. You just need to add those methods to `ServiceCollectionExtensions` and then make them register your real and design time services.

> thanks anyway this guide has helped me to understand DI very much.

Glad I could help. I put a lot of work into writing it!

**Reply by zangetsudo** [reply]
i finally git Design to work:

```xml
<UserControl
    x:Class="AvaloniaApplication2.Views.SettingsPageView"
    xmlns="https://github.com/avaloniaui"
    xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
    xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
    xmlns:markup="using:AvaloniaApplication2.Extension"
    xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
    xmlns:vm="using:AvaloniaApplication2.ViewModels"
    Design.DataContext="{markup:DesignLocator {x:Type vm:SettingsPageViewModel}}"
    d:DesignHeight="450"
    d:DesignWidth="800"
    x:DataType="vm:SettingsPageViewModel"
    mc:Ignorable="d">
```

thanks for `DesignLocator` from above post:
```cs
    public class DesignLocator : MarkupExtension
    {
        #region Public properties
        public Type Type { get; init; }
        #endregion

        #region Public methods

        public DesignLocator(Type type)
        {
            Type = type;
        }

        public override object ProvideValue(IServiceProvider serviceProvider)
        {
            if (!Design.IsDesignMode)
            {
                throw new InvalidOperationException();
            }
            
            return ServiceLocator.GetRequiredKeyedService<ViewModelBase>(Type.Name) ?? new object();
        }
        #endregion
    }
```

but cannot operator smoothly, after change the UI, it will not update until seconds. any ideas ?

**Reply by thevortexcloud** [reply]
> but cannot operator smoothly, after change the UI, it will not update until seconds. any ideas ?

I am going to need more context. 

* What are your services doing? 
* How are your services registered?
* What are you injecting into your view models?

Most DI systems are lazily initialised. Meaning services wont be created until they are actually needed.

>  ServiceLocator.GetRequiredKeyedService<ViewModelBase>(Type.Name) ?? new object();

This code also does not make sense to me. A required service will throw if it can't be found, meaning the null fall back wont ever do anything (unless you for some reason registered a null service). A keyed service is basically just a named service so you can get a specific version of a service implementation. But this code for some reason uses the type name as the key? There is no reason at all to do this. Just use the standard type retrieval system like I did in my original example.

**Comment by agalanrosa** [comment]
Can someone please share a complete implementation about this, I'm not able to get it working, I have the @thevortexcloud implementation and Added the `Design.DataContext={markup:DesignResolve {x:Type vm:MainViewModel}}` to my view `.xaml` file but it's not working.

cc @jp2masa 

Now I'm getting the same error as @leowest77 

LastUpdate: It's wokring now, I had to add constructor with the Type field, and add the markup namespace to the view

**Reply by thevortexcloud** [reply]
As far as I am aware, my code should work as is. It however has been a long time since I tried it.
